BaseClasses.py

```diff
@@ -681,19 +681,108 @@ def can_reach_region(self, spot: str, player: int) -> bool:
         return self.multiworld.get_region(spot, player).can_reach(self)
 
     def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None) -> None:
-        if locations is None:
-            locations = self.multiworld.get_filled_locations()
-        reachable_events = True
         # since the loop has a good chance to run more than once, only filter the events once
-        locations = {location for location in locations if location.advancement and location not in self.events and
-                     not key_only or getattr(location.item, "locked_dungeon_item", False)}
-        while reachable_events:
-            reachable_events = {location for location in locations if location.can_reach(self)}
-            locations -= reachable_events
-            for event in reachable_events:
-                self.events.add(event)
-                assert isinstance(event.item, Item), "tried to collect Event with no Item"
-                self.collect(event.item, True, event)
+        if key_only:
+            def event_filter(location: Location):
+                return (location.advancement
+                        and location not in self.events
+                        and getattr(location.item, "locked_dungeon_item", False))
+        else:
+            def event_filter(location: Location):
+                return location.advancement and location not in self.events
+
+        events_per_player: List[Tuple[int, List[Location]]]
+        if locations is None:
+            # `self.multiworld.get_filled_locations(player)` is avoided because it first iterates into a list and also
+            # because `location.advancement` in `event_filter` also checks for `location.item is not None`.
+            events_per_player = [(player, list(filter(event_filter, locations_dict.values())))
+                                 for player, locations_dict in self.multiworld.regions.location_cache.items()]
+            # Remove empty lists to reduce the length of `events_per_player`, so that it is quicker to iterate.
+            events_per_player = [(player, locations) for player, locations in events_per_player if locations]
+        else:
+            # Can't iterate self.multiworld.player_ids because it doesn't include the extra player IDs used for item
+            # links, so iterate the keys of the location_cache instead.
+            events_per_player_dict = {}
```

Comment:
```suggestion
            events_per_player_dict = default_dict(list)
```
or where you do the check further down just replace it with `events_per_player_dict.setdefault(player, []).append(locations)`

Reply:
I can try them. From my experience, I expect them to perform worse, however, if the difference is negligible, then picking one of them will definitely make the code more readable imo, which is always good.